### ISSUE TITLE
API does not allow you to execute a transaction

### ISSUE THREAD

**Issue by josephg** [agisboye/SwiftLMDB#7]
As far as I can tell, there's no way to run an LMDB transaction through SwiftLMDB. I need transactions for an app I'm building.

It looks like the code is all there - just the useful Transaction.init function isn't exposed. Please either make it public or provide another API through which I can run transactions.

**Comment by agisboye** [comment]
Hi @josephg,

Its not just a matter of exposing the initializer. Take a look at how the get, put and delete methods are [implemented](https://github.com/agisboye/SwiftLMDB/blob/4301eab886d6ef539e3f695fccddff3ec482b1c6/Sources/Database.swift#L132). They use a transaction internally. 

You would have to change a couple of things. I think the best approach would be to pass in get, put and delete closures to the transaction block. When you are done, you can either commit or abort.
If any of the calls inside the transaction block fail, the whole transaction is aborted.

What do you think of this?

```swift

let database: Database // ...


try database.transaction { get, put, delete -> Transaction.Result in

    // Get a value
    let value = try get(type: String.self, forKey: "key1")

    // Set a value
    try put(value: "Hello world!", forKey: "key1")

    // Delete a value
    try delete(forKey: "key1")

    return .commit // or .abort
            
}
```

**Comment by josephg** [comment]
That can work, though its little awkward pulling data out of the closure like that. You'd need to do something like this:

```swift
var foo: String?
var bar: Int?
try database.transaction { get, put, delete -> Transaction.Result in
    foo = try get(type: String.self, forKey: "fookey")
    bar = try get(type: Int.self, forKey: "barkey")
    return .commit
}
```

**Comment by agisboye** [comment]
You're absolutely right. I intend to keep the current shorthand methods on
Database, so for pulling out data you can still use Database.get. No need
to fire up a transaction (the method will do it for you).

Do you wanna give it a go? I'd be happy to merge in a PR.

**Comment by josephg** [comment]
You can, but for my use case I'll need to group multiple fetches in transactions anyway.

I'm not sure if I'll submit a patch - I'm at about week 2 using Swift. But I'll have a play with it. :)

**Comment by agisboye** [comment]
It'll be a great way to learn. :-)
Let me know if you have questions.

**Comment by josephg** [comment]
Um, lots :)

I've basically been rewriting a lot of the code. I wanna get the tests passing again before I push it though.

Um, doesn't this leak memory in `Database.get()`?

```swift
let dataPointer = UnsafeMutablePointer<MDB_val>.allocate(capacity: 1)
```

.. there's no corresponding release.

**Comment by josephg** [comment]
I'd love it if you could take a look at this patch to let me know if I'm doing anything dumb.

https://github.com/josephg/SwiftLMDB/commit/fee20dd99e03e956a562c7f321db3f5798efe943

I'm not sure if you're interested in merging it back - its ok if you aren't; I ended up making a lot more changes than I expected and its small enough to be easy enough to maintain. I'm going to see if I can add cursor support next.

**Comment by agisboye** [comment]
According to the [docs](http://104.237.133.194/doc/group__mdb.html#ga8bf10cd91d3f3a83a34d04ce6b07992d):
> The memory pointed to by the returned values is owned by the database. The caller need not dispose of the memory, and may not modify it in any way. 

That originally led me to believe we don't need to have a release. Reading the documentation again, I'm thinking we need to release the MDB_val, which itself will release the memory that it points to. I'll have to test it. Good catch. 👍 

Regarding the patch:
It's a bit more extensive than I had hoped. I'm not a fan of replacing all of the Swift types with UnsafeMutablePointer. We could just use the C implementation directly then. 😉  It makes it difficult to maintain, and sometimes copying values is actually desirable.

I think I'll try my hand to show what I had in mind. We can compare notes then.

**Comment by josephg** [comment]
The way you've written that code as I understand it you're allocating memory for the MDB_val on the heap. You don't need to do that - its a tiny struct (pointer and size). Just stack-allocate.

```swift
var dataVal = MDB_val()
let result = mdb_get(self.handle, dbi, &keyVal, &dataVal)
```

As for the patch, I would have said the same thing - sorry about that :) . Once I started making changes I realised there were lots of things like that which I would like to change, like that heap allocated MDB_val. Little unnecessary memory copies and allocations like that make my programmer-OCD itch.

For my use I'm still trying to figure out what I want - I'm porting some server code from node which makes very heavy use of cursors and transactions, so I need to either wrap all that stuff or just use the C API directly. Its fun playing with expressing lmdb through swift, but I'm not sure how much of a difference the ergonomics will actually make. (Although I'll definitely steal the error code wrapping either way. And probably some variant of the type coercion code.)
